src/control/lib/spdk/src/nvme_control.c

```diff
@@ -65,121 +71,204 @@ get_dev_health_logs(struct spdk_nvme_ctrlr *ctrlr,
-		return ret;
-	}
+static struct ret_t *
+wipe(char *ctrlr_pci_addr)
+{
+	struct ctrlr_entry	*centry;
+	struct ns_entry		*nentry;
+	struct ret_t		*ret;
+	struct wipe_sequence	 sequence;
+	int			 rc;
 
-	if (spdk_nvme_ctrlr_get_next_active_ns(sequence.ctrlr, ns_id) != 0) {
-		/* unexpected, don't wipe */
-		snprintf(ret->err, sizeof(ret->err),
-			"expected a single active namespace, got multiple");
-		ret->rc = -NVMEC_ERR_MULTIPLE_ACTIVE_NS;
-		return ret;
-	}
+	ret = init_ret(0);
 
-	sequence.ns = spdk_nvme_ctrlr_get_ns(sequence.ctrlr, ns_id);
-	if (!sequence.ns) {
-		snprintf(ret->err, sizeof(ret->err), "null ns");
-		ret->rc = -NVMEC_ERR_NULL_NS;
+	rc = get_controller(&centry, ctrlr_pci_addr);
+	if (rc != 0) {
+		snprintf(ret->info, sizeof(ret->info), "get_controller()\n");
+		ret->rc = rc;
 		return ret;
 	}
 
-	sequence.qpair = spdk_nvme_ctrlr_alloc_io_qpair(sequence.ctrlr, NULL,
-							0);
-	if (sequence.qpair == NULL) {
-		snprintf(ret->err, sizeof(ret->err),
-			"spdk_nvme_ctrlr_alloc_io_qpair() failed");
-		ret->rc = -NVMEC_ERR_ALLOC_IO_QPAIR;
-		return ret;
-	}
+	nentry = centry->nss;
+	while (nentry != NULL) {
+		nentry->qpair = spdk_nvme_ctrlr_alloc_io_qpair(nentry->ctrlr,
+							       NULL, 0);
+		if (nentry->qpair == NULL) {
+			snprintf(ret->info, sizeof(ret->info),
+				 "spdk_nvme_ctrlr_alloc_io_qpair()\n");
+			ret->rc = -1;
+			return ret;
+		}
 
-	sequence.using_cmb_io = 1;
-	sequence.buf = spdk_nvme_ctrlr_alloc_cmb_io_buffer(sequence.ctrlr,
-							   0x1000);
-	if (sequence.buf == NULL) {
-		sequence.using_cmb_io = 0;
 		sequence.buf = spdk_zmalloc(0x1000, 0x1000, NULL,
 					    SPDK_ENV_SOCKET_ID_ANY,
 					    SPDK_MALLOC_DMA);
+		if (sequence.buf == NULL) {
+			snprintf(ret->info, sizeof(ret->info),
+				 "spdk_zmalloc()\n");
+			ret->rc = -1;
+			spdk_nvme_ctrlr_free_io_qpair(nentry->qpair);
+			return ret;
+		}
+		sequence.is_completed = 0;
+		sequence.ns_entry = nentry;
+
+		rc = spdk_nvme_ns_cmd_write(nentry->ns, nentry->qpair,
+					    sequence.buf, 0, /* LBA start */
+					    1, /* number of LBAs */
+					    write_complete, &sequence, 0);
+		if (rc != 0) {
+			snprintf(ret->info, sizeof(ret->info),
+				 "spdk_nvme_ns_cmd_write() (%d)\n", rc);
+			ret->rc = -1;
+			spdk_free(sequence.buf);
+			spdk_nvme_ctrlr_free_io_qpair(nentry->qpair);
+			return ret;
+		}
+
+		while (!sequence.is_completed) {
+			rc = spdk_nvme_qpair_process_completions(nentry->qpair,
+								 0);
+			if (rc < 0) {
+				fprintf(stderr,
+					"process completions returns %d\n", rc);
+				break;
+			}
+		}
+
+		if (sequence.is_completed != 1) {
+			snprintf(ret->info, sizeof(ret->info),
+				 "spdk_nvme_ns_cmd_write() callback\n");
+			ret->rc = -1;
+			spdk_free(sequence.buf);
+			spdk_nvme_ctrlr_free_io_qpair(nentry->qpair);
+			return ret;
+		}
+
+		snprintf(ret->info + sizeof(ret->info),
+			 sizeof(TEXTLEN) - strlen(ret->info), "%d ",
+			 spdk_nvme_ns_get_id(nentry->ns));
+
+		spdk_free(sequence.buf);
```

Comment:
Is anyone actually checking the content of sequence.buf? If not could the field be removed entirely, and the read fully handled at a lower level?

Reply:
the read back seems to be mandatory in order to get the cmd_write working as expected, cmd_write_zeroes does not have a respective supported opcode on the GCP NVMe implementation so we have to allocate this zeroed buffer to write/read

Reply:
It makes sense to allocate, but I'm wondering if it could be fully handled somewhere further down the stack, instead of keeping the buffer in the struct and returning it up to be freed here.

Reply:
done although not sure it adds significant value